web-client: Enable HD wallet key derivation

sisou opened the pull request:
Add classes/structs and methods to support HD wallet key derivation as was possible with the core-js library.

- [ ] Review of encryption/decryption methods by @paberr and/or @hrxi for best practices
- [x] Move of CryptoUtils/hashing methods into their own worker so they don't block the main thread

sisou committed:
web-client: Add CryptoUtils, Entropy, ExtendedPrivateKey and MnemonicUtils

The encryption is PoW-backward-compatible (using the Argon2d algorithm).

sisou committed:
web-client: Add Signature.create() method

sisou committed:
web-client: Rename toBytes/fromBytes => serialize/unserialize for signatures

sisou committed:
Create a new WASM bundle for CryptoUtils and run it in a worker

sisou committed:
Clean up web-client NodeJS example

sisou committed:
Move key derivation to JS lib

That way these classes can be tree-shaken and do not blow up WASM size when not used.

Move CryptoUtils.otpKdf function into an auto-spawning worker, so it doesn't block the main thread, as it usually takes a few seconds to run, other than the other methods on the class.

Eligioo commented on key-derivation/src/lib.rs:
```diff
@@ -5,8 +5,9 @@ use nimiq_hash::{hmac::*, sha512::Sha512Hash};
 use nimiq_keys::{Address, Ed25519PublicKey, PrivateKey};
 use nimiq_serde::Serialize;
 use regex::Regex;
+use serde::Deserialize;
```
Instead of using `serde::Deserialize`, shouldn't we use `nimiq_serde::Deserialize`?

Eligioo commented on mnemonic/src/lib.rs:
```diff
@@ -58,6 +63,106 @@ impl Entropy {
+        let rounds = 2u32.pow(rounds_log as u32);
+
+        match version {
+            1 => unimplemented!(),
```
Are we going to fully discard older versions?

sisou replied:
The older versions use a KDF OTP function that core-rs-albatross no longer implements.

Eligioo commented on web-client/example/README.md:
````diff
@@ -24,7 +24,7 @@ to build the WASM files. Then go into `example/node` and run
 ```
 
-to install the dependencies. Other package managers will also work, but the `preinstall` script uses `yarn`.
+to install the dependencies. Other package managers will also work.
````
Yarn isn't mandatory all together anymore?

sisou replied:
Not here, as I removed the preinstall script (below). So packages can be installed with any manager for the example.

Eligioo reviewed (commented):
Halfway through the review, already some comments.

Eligioo commented on web-client/src/primitives/es256_signature.rs:
```diff
@@ -17,24 +17,22 @@ impl ES256Signature {
     /// Throws when the byte array contains less than 64 bytes.
-    #[wasm_bindgen(js_name = fromBytes)]
-    pub fn deserialize(bytes: &[u8]) -> Result<ES256Signature, JsError> {
+    pub fn unserialize(bytes: &[u8]) -> Result<ES256Signature, JsError> {
```
I assume you switched from `deserialize` to `unserialize` to stay closer to the core-js PoW naming convention? 

Though throughout the rest of the Rust codebase, `deserialize` is the naming convention for deserializing bytes. And as the `web-client` is breaking anyway compared to the PoW core-js, imo this is the right time to use a different naming convention. Then I would be in favor of using `deserialize` over `unserialize`.

Eligioo commented on web-client/extras/launcher/cryptoutils-worker-proxy.ts:
```diff
@@ -0,0 +1,60 @@
+
+let remote: Remote<CryptoUtilsWorker> | undefined;
+
+// class CryptoUtilsProxy {}
```
Forgot to remove?

Eligioo commented on web-client/extras/lib/Secret.ts:
```diff
@@ -0,0 +1,154 @@
+import { SerialBuffer } from './SerialBuffer';
+import { Serializable } from './Serializable';
+
+abstract class Secret extends Serializable {
```
Would it be an idea to also port over the tests? For the `ExtendedPrivateKey`, `Secret`, etc.?

sisou replied:
I can try, but then we need a JS test-runner, too :/

sisou replied:
Added the relevant tests with `vitest` :muscle:

sisou committed:
Rebase fixes, review fixes

sisou committed:
Add unittests for web-client JS lib

Copied the relevant tests from core-js.

sisou committed:
Replace StringUtils.isMultiByte with StringUtils.isWellFormed

Wrapping String.prototype.isWellFormed or a replacement taken from the official polyfill.

sisou committed:
Update JS varint implementation to be the same as in Rust

sisou commented:
@jsdanielh Rebased and squashed a few earlier commit together.

Eligioo commented:
Has this been further reviewed?
